app/channel_category.go

```diff
@@ -144,14 +144,135 @@ func (a *App) UpdateSidebarCategoryOrder(userId, teamId string, categoryOrder []
+	for i, updatedCategory := range updatedCategories {
+		originalCategory := originalCategories[i]
+
+		if updatedCategory.Muted && !originalCategory.Muted {
+			channelsToMute = append(channelsToMute, updatedCategory.Channels...)
+		} else if !updatedCategory.Muted && originalCategory.Muted {
+			channelsToUnmute = append(channelsToUnmute, updatedCategory.Channels...)
+		}
+	}
+
+	// Mute any channels moved from an unmuted category into a muted one and vice versa
+	channelsDiff := diffChannelsBetweenCategories(updatedCategories, originalCategories)
+	if len(channelsDiff) != 0 {
+		makeCategoryMap := func(categories []*model.SidebarCategoryWithChannels) map[string]*model.SidebarCategoryWithChannels {
+			result := make(map[string]*model.SidebarCategoryWithChannels)
+			for _, category := range categories {
+				result[category.Id] = category
+			}
+
+			return result
+		}
+
+		updatedCategoriesById := makeCategoryMap(updatedCategories)
+		originalCategoriesById := makeCategoryMap(originalCategories)
+
+		for channelId, diff := range channelsDiff {
+			fromCategory := originalCategoriesById[diff.fromCategoryId]
+			toCategory := updatedCategoriesById[diff.toCategoryId]
+
+			if toCategory.Muted && !fromCategory.Muted {
+				channelsToMute = append(channelsToMute, channelId)
+			} else if !toCategory.Muted && fromCategory.Muted {
+				channelsToUnmute = append(channelsToUnmute, channelId)
+			}
+		}
+	}
+
+	for _, channelId := range channelsToMute {
+		_, err := a.setChannelMuted(channelId, userId, true)
+		if err != nil {
+			mlog.Error(
+				"Failed to mute channel to match category",
+				mlog.String("channel_id", channelId),
+				mlog.String("user_id", userId),
+				mlog.Err(err),
+			)
+		}
+	}
+
+	for _, channelId := range channelsToUnmute {
+		_, err := a.setChannelMuted(channelId, userId, false)
+		if err != nil {
+			mlog.Error(
+				"Failed to unmute channel to match category",
+				mlog.String("channel_id", channelId),
+				mlog.String("user_id", userId),
+				mlog.Err(err),
+			)
+		}
+	}
+}
+
+type categoryChannelDiff struct {
+	fromCategoryId string
+	toCategoryId   string
+}
+
+func diffChannelsBetweenCategories(updatedCategories []*model.SidebarCategoryWithChannels, originalCategories []*model.SidebarCategoryWithChannels) map[string]*categoryChannelDiff {
+	// mapChannelIdsToCategories returns a map of channel IDs to the IDs of the categories that they're a member of.
+	mapChannelIdsToCategories := func(categories []*model.SidebarCategoryWithChannels) map[string]string {
+		result := make(map[string]string)
+		for _, category := range categories {
+			for _, channelId := range category.Channels {
+				result[channelId] = category.Id
+			}
+		}
+
+		return result
+	}
+
+	updatedChannelIdsMap := mapChannelIdsToCategories(updatedCategories)
+	originalChannelIdsMap := mapChannelIdsToCategories(originalCategories)
+
+	// Check for any channels that have changed categories. Note that we don't worry about any channels that have moved
```

Comment:
> Note that we don't worry about any channels that have moved outside of these categories

Curious to know why is this. Is it because the webapp always correctly uses the API? Or is there another reason?

Reply:
Yeah, when moving a channel from category A to B, the webapp will always call the API with both categories, but you could theoretically move the channel by just updating B with the new ID. That case will fail to mute the channel, but checking that would require extra database hits to find which category the channel was originally in before doing the updated, so keeping track of that got really complicated.

Since it should work with how we currently use the API, I wasn't too worried about it working incorrectly in an edge case at the moment. I left the note to explain that in case someone ever runs into it in the future.